homeassistant/components/knx/config_flow.py

```diff
@@ -0,0 +1,273 @@
+"""Config flow for KNX."""
+from __future__ import annotations
+
+from collections import OrderedDict
+from typing import Final
+
+import voluptuous as vol
+from xknx import XKNX
+from xknx.io import DEFAULT_MCAST_GRP, DEFAULT_MCAST_PORT
+from xknx.io.gateway_scanner import GatewayDescriptor, GatewayScanner
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_PORT
+from homeassistant.data_entry_flow import FlowResult
+
+from .const import (
+    CONF_KNX_AUTOMATIC,
+    CONF_KNX_CONNECTION_TYPE,
+    CONF_KNX_INDIVIDUAL_ADDRESS,
+    CONF_KNX_ROUTING,
+    CONF_KNX_TUNNELING,
+    DOMAIN,
+)
+from .schema import ConnectionSchema
+
+CONF_KNX_GATEWAY: Final = "gateway"
+
+
+class FlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a KNX config flow."""
+
+    VERSION = 1
+
+    _tunnels: list = []
+
+    async def async_step_user(self, user_input: dict | None = None) -> FlowResult:
+        """Handle a flow initialized by the user."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+
+        self._tunnels = []
+        return await self.async_step_type()
+
+    async def async_step_type(self, user_input: dict | None = None) -> FlowResult:
+        """Handle connection type configuration."""
+        errors: dict = {}
+        supported_connection_types = [CONF_KNX_TUNNELING, CONF_KNX_ROUTING]
+
+        if user_input is None:
+            gateways = await scan_for_gateways()
+
+            if gateways:
+                supported_connection_types.insert(0, CONF_KNX_AUTOMATIC)
+                for gateway in gateways:
+                    if gateway.supports_tunnelling:
+                        self._tunnels.append(gateway)
+
+        if user_input is not None:
+            connection_type = user_input[CONF_KNX_CONNECTION_TYPE]
+            if connection_type == CONF_KNX_AUTOMATIC:
+                return self.async_create_entry(
+                    title=CONF_KNX_AUTOMATIC, data=user_input
+                )
+
+            if connection_type == CONF_KNX_ROUTING:
+                return await self.async_step_routing()
+
+            if connection_type == CONF_KNX_TUNNELING and self._tunnels:
+                return await self.async_step_tunnel()
+
+            return await self.async_step_manual_tunnel(user_input)
+
+        fields = OrderedDict()
+        fields[vol.Required(CONF_KNX_CONNECTION_TYPE)] = vol.In(
+            supported_connection_types
+        )
```

Comment:
All dictionaries are ordered since Python 3.6
```suggestion
        fields = {
            vol.Required(CONF_KNX_CONNECTION_TYPE)] = vol.In(
                supported_connection_types
            )
        }
```